File: Python Snippets/merge_sort.py
from random import randint

def merge_sort(nums):
  if len(nums) <= 1:
    return nums

  mid = len(nums) // 2

  left = merge_sort(nums[:mid])
  right = merge_sort(nums[mid:])

  return merge(left, right)

def merge(left, right):
  merged_array = []
  i = j = 0

  while i < len(left) and j < len(right):
    if left[i] <= right[j]:
      merged_array.append(left[i])
      i += 1
    else:
      merged_array.append(right[j])
      j += 1

  merged_array += left[i:]
  merged_array += right[j:]

  return merged_array

unsorted_array = [randint(-10, 10) for i in range(10)]
print(f"Unsorted Array = {unsorted_array}")
print(f"Sorted Array = {merge_sort(unsorted_array)}")

File: R Snippets/merge_sort.r
merge_sort <- function(nums) {
  if (length(nums) <= 1) {
    return (nums)
  }

  size <- length(nums)
  mid <- size %/% 2

  left <- merge_sort(nums[1:mid])
  right <- merge_sort(nums[(mid + 1):size])

  return (merge(left, right))
}

merge <- function(left, right) {
  i <- 1
  j <- 1
  merged <- c()

  while (i < length(left) + 1 & j < length(right) + 1) {
    if (left[i] <= right[j]) {
      merged <- append(merged, left[i])
      i <- i + 1
    } else {
      merged <- append(merged, right[j])
      j <- j + 1
    }
  }

  if (i < length(left) + 1) {
    merged <- append(merged, left[i:length(left)])
  }
  if (j < length(right) + 1) {
    merged <- append(merged, right[j:length(right)])
  }

  return (merged)
}

sample_numbers <- sample(-10:10, 10)
sorted_numbers <- merge_sort(sample_numbers)
cat("Unsorted Array =", sample_numbers, "\n")
cat("Sorted Array =", sorted_numbers, "\n")